topics/isthisamock.md

```diff
@@ -0,0 +1,264 @@
+    assert(c.Runtime(), IsZero())
+}
+```
+
+The dummy monitor is passed because the function signature requires it.
+The test itself is not expected to use it.
+
+A `Dummy` does *nothing*. It acts as a placeholder.
+If its methods return something, it will be usually the default value
+(e.g. `nil`, `0`).
+
+Moreover, if someone does try to use it in the production code, it
+should probably blow up.
+
+## Stub
+
+```go
+// Production
+func (c *Computer) Start() error {
+    if err := c.monitor.Display("Starting the computer"); err != nil {
+        return err
+    }
+    // Do the needed to start a computer.
+
+    return nil
+}
+
+// Test Stub
+type StubMonitorDisplayFails struct {
+    err error
+}
+func (s StubMonitorDisplayFails) Display(fmt string, args ...any) {
+	return s.err 
+}
+
+// Test
+func testComputerIsStatringWithDisplayFailure() {
+    monitor := StubMonitorDisplayFails{errors.New("an error")}
+    c := NewComputer(monitor)
+    assert(c.Start(), hasError("an error"))
+}
+```
+
+The stub returns or sets a specific value, allowing us to check the
+tested unit in that condition.
+
+The example above defines a stub with a specific error and expects
+to see the error returning when calling `Start()`.
+
+## Spy
+
+```go
+// Same production code as before
+
+// Test Spy
+type SpyMonitor struct {
+    displayData string
+}
+func (s *SpyMonitor) Display(fmt string, args ...any) {
+    s.displayData = fmt.Sprintf(fmt, args...)
+}
+
+// Test
+func testComputerIsStatring() {
+    monitor := SpyMonitor{}
+    c := NewComputer(monitor)
+    assert(c.Start(), noError())
+    assert(monitor.displayData, isEqual("Starting the computer"))
+}
+```
+
+Spy record calls to it.
+It can record how many times it was called or with which arguments.
+
+But overusing them is dangerous, it makes tests fragile and dependent
+on implementation of the production code (and not on their API).
+
+## Mock
+
+```go
+// Same production code as before
+
+// Test Stub
+type MockMonitor struct {
+    displayData         string
+    expectedDisplayData string
+}
+func (m *MockMonitor) Display(fmt string, args ...any) {
+    m.data = fmt.Sprintf(fmt, args...)
+}
+
+func (m *MockLog) Assert() {
+    assert(m.displayData, isEqual(m.expectedDisplayData))
+}
+
+// Test
+func testComputerIsStatring() {
+    monitor := MockDisplay{expectedDisplayData: "Started the computer"}
+    c := NewComputer(monitor)
+    assert(c.Start(), noError())
+    monitor.Assert()
+}
+```
+
+A mock is set up in advance with the required data, including the
+details of what it expects at the behavior level.
+
+It verifies and perform the assertion by itself.
+
+Mocks have become popular because tooling have been built to generate
+them for a given interface automatically.
+
+Its disadvantage is similar to spies.
+It requires the setup stage and mixes expectations with assertions and
+when used with autogenerated tooling, it usually does a lot more than it
+is needed.
+
+## Fake
```

Comment:
Here is a real life example of a fake: https://github.com/kubevirt/kubevirt/blob/c6b4ce6ea8c3ae08b4ffb83b815b7727e854477e/pkg/network/driver/netlink/fake/fake.go#L28

It mimics this code: https://github.com/kubevirt/kubevirt/tree/c6b4ce6ea8c3ae08b4ffb83b815b7727e854477e/pkg/network/driver/netlink